### ISSUE TITLE
Unreal Engine GSDK plugin memory leak in StandingBy

### ISSUE THREAD

**Issue by blastodd** [PlayFab/gsdk#108]
**Describe the bug**
Unreal engine dedicated server using GSDK plugin will slowly consume more and more RAM over time. 

**To Reproduce**
1. Add PlayFabGSDK to ThirdPersonMP project
2. Build linux dedicated server using Docker
3. Run LocalMultiplayerAgent on the server image with NumHeartBeatsForActivateResponse set to a large number to simulate StandingBy
4. See Docker Container Stats Memory usage

**Expected behavior**
Memory Usage should not rise continuously

**Screenshots**
![memoryleak](https://user-images.githubusercontent.com/64591018/143948043-28b63e65-774a-4f8b-885d-e64591459041.PNG)

**Desktop (please complete the following information):**
 - OS: Linux
 - Latest GSDK for Unreal Engine Plugin

**Additional context*
I am running two servers on a playfab VM, and in the first part (before 12:00) the servers are in standing by, and you can observe they take turns eating up memory, crashing, and reopening.
After 12 I requested both servers to become active, closed them to reset their memory usage, and then request them both again, which sets them in the Active State. As you can see from that point forward, memory availability is stable.
Until 15:00 when the servers were closed and new ones were back in StandingBy state.

**Comment by dgkanatsios** [comment]
FYI @stkrwork 
Thanks for reporting that! The screenshots you pasted are from a PlayFab VM, correct? Any data from the machine you faced the issue with LocalMultiplayerAgent? Any chance you can do a process dump to see what's eating memory?

I'm probably missing something, but I don't see any issue with the memory in standingBy. Servers on MPS never "re-open". Once the game server process exits (either gracefully or by a crash) we cleanup the docker container, archive and upload logs and create a new one in its turn. In your case, containers are created, memory goes to ~50%, containers crash, memory goes to 50% again and so on and so forth. Sorry if I don't see the issue, can you please clarify?

Here's also a helpful link on how Linux OS manages memory, it might provide additional context - https://www.linuxatemyram.com/

**Comment by pgilmorepf** [comment]
We will investigate. I expect to have some result no sooner than January. If you can provide info about what specifically may be leaking, this would be helpful to speed up the process.

**Comment by blastodd** [comment]
> Thanks for reporting that! The screenshots you pasted are from a PlayFab VM, correct? Any data from the machine you faced the issue with LocalMultiplayerAgent? Any chance you can do a process dump to see what's eating memory?

Yes, the screenshots are from a Playfab VM ( Dasv4 )
I will try to get more data.

> I'm probably missing something, but I don't see any issue with the memory in standingBy. Servers on MPS never "re-open". Once the game server process exits (either gracefully or by a crash) we cleanup the docker container, archive and upload logs and create a new one in its turn. In your case, containers are created, memory goes to ~50%, containers crash, memory goes to 50% again and so on and so forth. Sorry if I don't see the issue, can you please clarify?

I was under the impression that if the server goes to Active state it will retain the memory amount, but it seems to not be the case. Yes, the containers recycle, but let's say there will be 8 active servers and 2 servers StandingBy that will consume memory, what are the chances of an active server crashing ?


> We will investigate. I expect to have some result no sooner than January. If you can provide info about what specifically may be leaking, this would be helpful to speed up the process.

I will try to get more info.

**Comment by playvalve-teamcity** [comment]
@dgkanatsios - to quote you:
In your case, containers are created, memory goes to ~50%, **containers crash**, memory goes to 50% again and so on and so forth. Sorry if I don't see the issue, can you please clarify?

I really don't think containers should crash.
And as @blastodd  explained, the memory is **NOT** behaving that way **unless you add GSDK** in the project.
Sorry but then I don't see how this is not a GSDK issue nor I see how you can't see the issue with that.

You're not getting away just saying that "that's how linux manages memory" quick fix.

**Comment by dgkanatsios** [comment]
thanks both for responding. So, from what I understand, the issue is not about a memory leak but about the containers crashing when transitioning to Active state. You mentioned that you ran your game server with LocalMultiplayerAgent, correct? Does the crash occur there? 

When the game server is running on PlayFab, we collect logs from exiting game servers which you can retrieve via Game Manager, check for details [here](https://docs.microsoft.com/en-us/gaming/playfab/features/multiplayer/servers/archiving-and-retrieving-multiplayer-server-logs).

**Comment by blastodd** [comment]
> thanks both for responding. So, from what I understand, the issue is not about a memory leak but about the containers crashing when transitioning to Active state. You mentioned that you ran your game server with LocalMultiplayerAgent, correct? Does the crash occur there?

I didn't say the containers crash when transitioning to Active state.
What I said is this: we know that standing by servers will crash and reopen, because they run out of memory. But let's say we'll have 8 active servers ( that are already active, not transitioning to active ), and 2 standing by servers that do the crashing, reopening. What are the chances of an active server running out of memory and crashing, because the 2 standing by servers are consuming it ?

This is my issue, or rather my concern. From my point of view, however we spin it, it's not an ok behavior.

And regarding LocalMultiplayerAgent, I used it to observe memory going up, and do the tests on a new clean project, and as far as I know it can create only one container.

At any rate, I will try and get more data.

**Comment by dgkanatsios** [comment]
Sorry, I misunderstood. However, are you 100% that the servers are crashing because the system is out of memory? Where did you find logs that this happens?

**Comment by blastodd** [comment]
> In your case, containers are created, memory goes to ~50%, containers crash, memory goes to 50% again and so on and so forth.
> Sorry, I misunderstood. However, are you 100% that the servers are crashing because the system is out of memory? Where did you find logs that this happens?

Just to touch up a bit on this/clarify, the graph in the first post shows Memory Availability, which means that when it goes towards 0%, it means it's using the full VM memory, and when it's going back up to 50%, a container was stopped and recreated.

**Comment by dgkanatsios** [comment]
gotcha - yeah, the servers are *maybe* going down/crashing because of unavailable memory, I'd just want to make 100% sure that this is the case. If you run a smaller number of servers on the VM, do they crash?

**Comment by blastodd** [comment]
Currently there are only 2 servers on Standing By on the VM. I don't really know what to show, or what tests to do.
![image](https://user-images.githubusercontent.com/64591018/144114777-e93891bf-3e1e-4471-b3d5-f7b882a7a76f.png)
![image](https://user-images.githubusercontent.com/64591018/144114809-9e07444e-a74c-4997-bf5d-4a8a7d324ff0.png)

Active servers might not be affected, they might not crash, but still, there's a possibility.

**Comment by dgkanatsios** [comment]
Thank you. Yeah let’s focus on fixing the standingBy crash issue. So two servers on the VM crash,right? What if you create just one? Does it crash?

**Comment by blastodd** [comment]
> Thank you. Yeah let’s focus on fixing the standingBy crash issue. So two servers on the VM crash,right? What if you create just one? Does it crash?

There are two servers, but they don't both crash at the same time as you can see in the picture, only the largest one crashes. They take turns in a way. If there'd be only one, it would take a longer time, but eventually that one will restart as well.

**Comment by dgkanatsios** [comment]
- first of all, the ~`00% CPU usage is not healthy. Is there anything going on at the game server process level? This should be investigated
- are the archived logs showing anything? https://docs.microsoft.com/en-us/gaming/playfab/features/multiplayer/servers/archiving-and-retrieving-multiplayer-server-logs
- maybe you check Docker daemon logs? ```sudo journalctl -fu docker.service```
- ```sudo journalctl``` should give system logs, maybe there is something there
- ```docker logs -f containerName``` will give you streaming logs for the container, maybe you can use that to see what it outputs before it crashes (even though these logs should be visible by the first bullet)
- last but not least, you should profile your Unreal application. I'm not familiar with the Unreal, but this link contains some information on how to do it https://docs.unrealengine.com/4.27/en-US/TestingAndOptimization/PerformanceAndProfiling/Profiler/

**Comment by pgilmorepf** [comment]
# Disclaimer:

UE GSDK is still in pre-release preview. We have recently removed this disclaimer from the documentation in preparation for upcoming release and announcement. This bug came in just as we were doing final testing internally for that release announcement. All our Windows testing is done, and we're just starting Linux testing now, so it's relevant that the issue appears to be Linux only.

# What I need:

Logs, diagnostics, confirmation of process crash, specific info about the cause

# What I have:

* Memory graph report, indicating periodic 100% memory consumption
* Statement that it only happens when GSDK is present (Unconfirmed, but we'll assume true)
* Strong evidence that this does not occur in our internal Windows testing

# How I intend to proceed:

For now, we'll just look into our memory allocation/cleanup, looking for obvious causes. It's likely this will be slow (hence my earlier statement: "No sooner than January"). If we can get some logs, diagnostics, or concrete information about the underlying cause, we can probably have a more optimistic timetable.

# Other Notes:

This Unreal plugin is customer submitted. We're vetting the code and hosting it, but we're not the authors of it. We are learning it as we go, and that process is still happening. Thank you for your understanding. This is still an active investigation, so please be patient and kind.

**Comment by stkrwork** [comment]
Hey, sorry to hear you are having issues with the plugin. One thing that came to mind for me was if you are logging a lot while the game is in StandBy? The reason why I am asking is because while the game thread is locked, the log buffer is not flushed until unlocking.

Are you logging a message in the GSDK health check callback?

**Comment by blastodd** [comment]
> Hey, sorry to hear you are having issues with the plugin. One thing that came to mind for me was if you are logging a lot while the game is in StandBy? The reason why I am asking is because while the game thread is locked, the log buffer is not flushed until unlocking.
> 
> Are you logging a message in the GSDK health check callback?

Thanks for all the work put in.
No logging in the health check callback.

As I said in the first post, I replicated this with a new project based on ThirdPersonMP.
Just added GSDK to it, created a linux docker image using Unreal Containers by Adam Rehn.

For now I will wait for your conclusions/internal testing, but at the same time I will try and get more data.
Maybe it's just me and I am doing something wrong.

**Comment by pgilmorepf** [comment]
@blastodd, I know this is an oddly specific request, but despite your original report, your statements lack an absolute statement verifying that the server processes are actually crashing. I could think of a lot of ways/reasons why a process might consume all available memory, and then... GC or something, on the first failed allocation. And after that, never actually crash. Please keep in mind, I'm not saying it is or isn't, I just want a confirmation one way or the other, about whether the process crashes with an out-of-memory error, or recovers somehow. This would significantly affect how we proceed with our investigation.

Thanks.

**Comment by nassosterz-ms** [comment]
We would like to provide an update on this issue.

1.	We have replicated the issue on both the **Playfab VM** and using **LocalMultiPlayerAgent**. Thank you for bringing this up.
2.	We have circled down on what might be causing the constant memory consumption while the server is on **StandBy** mode.
3.	We are currently investigating potential solutions to this issue. We hope to have a concrete solution by the beginning of the new year.
